src/android.ts

```diff
@@ -318,47 +327,106 @@ export const androidApi = <RunTarget extends SupportedRunTarget<'android'>>(
+                expansions?: { file: string; install_location: string; install_path: string }[];
+                split_apks?: { file: string }[];
+            };
+            const xapk = await open(apkPath);
+            await getFileFromZip(xapk, 'manifest.json').then(async (manifest) => {
+                if (!manifest) throw new Error('Failed to install app: manifest.json not found in XAPK.');
+                const manifestString = await new Promise<string>((resolve) => {
+                    let result = '';
+                    manifest.on('data', (chunk) => (result += chunk.toString()));
+                    manifest.on('end', () => resolve(result));
+                });
+                const manifestJson: xapkManifest = JSON.parse(manifestString);
+
+                const expansionFileNames = manifestJson.expansions?.map((expansion) => expansion.file);
+                const apkFileNames = manifestJson.split_apks?.map((apk) => apk.file);
+                const tmpApks: string[] = [];
+                const externalStorageDir = (await execa('adb', ['shell', 'echo', '$EXTERNAL_STORAGE'])).stdout;
+
+                if (expansionFileNames?.length === 0) {
+                    await this._internal.requireRoot('writing to external storage in installApp');
+                }
+
+                await forEachInZip(xapk, async (entry, zipFile) => {
+                    if (apkFileNames?.includes(entry.fileName)) {
+                        await tmpFileFromZipEntry(zipFile, entry, 'apk').then((tmpFile) => void tmpApks.push(tmpFile));
+                    } else if (expansionFileNames?.includes(entry.fileName)) {
+                        const expansion = manifestJson.expansions?.find((exp) => exp.file === entry.fileName);
+                        if (expansion && expansion.install_location === 'EXTERNAL_STORAGE')
+                            // Since there doesn't seem to be any public schema of XAPKs and Google explicitly says that
+                            // extension files are supposed to be stored in the external storage, hardcoding this is the
+                            // only way to handle this
+                            // (https://github.com/tweaselORG/appstraction/issues/63#issuecomment-1514822176).
+                            return tmpFileFromZipEntry(zipFile, entry).then(async (tmpFile) => {
+                                await execa('adb', [
+                                    'push',
+                                    tmpFile,
+                                    `${externalStorageDir}/${expansion.install_path}`,
+                                ]);
+                                await rm(tmpFile);
+                            });
+                        throw new Error('Failed to install app: Invalid expansion file declaration.');
+                    }
+                }).then(xapk.close);
+
+                if (tmpApks.length === 0) throw new Error('Failed to install app: No split apks found in XAPK.');
```

Comment:
Yes, that would be ideal. But that would require looping through the whole zip again, which could take quite long, so this is a performance improvement. However, right now we also don't do any cleanup. I guess there are two options: either loop through the zip twice and first install and then push the obbs, or try to install each APK during the loop. Depending on the order of the files in the zip, this might still lead to obbs being pushed.

I want to push the obbs rights away, because they tend to be quite large and I don't want to keep them clogging up the cache for too long. So I can not just keep a list of them to later push if the app installed.

Reply:
> Yes, that would be ideal. But that would require looping through the whole zip again

Oh, I see. Yeah, that's fine then. Let's leave it as is.